src/emulator/Emulator.js

```diff
@@ -251,32 +281,41 @@ class EmulatorImpl {
         this.warnMempool()
 
         if (!tx.isValidSlot(BigInt(this.currentSlot))) {
-            throw new Error(
+            // TODO: add isExpired or isPremature, when other preferred clients also have positive detection
+            //  of these details
+            throw new SubmissionExpiryError(
                 `tx invalid (slot out of range, ${this.currentSlot} not in ${tx.body.getValidityTimeRange(this.parametersSync).toString()})`
             )
         }
 
         // make sure that each input exists
-        if (
-            !tx.body.inputs.every(
-                (input) => input.id.toString() in this._allUtxos
-            )
-        ) {
-            throw new Error("some inputs don't exist")
+        for (const input of tx.body.inputs) {
+            if (!(input.id.toString() in this._allUtxos)) {
```

Comment:
I think it might be cleaner to use the newly created `hasUtxo()` method instead.

Reply:
Yup, looks great that way.  done.